generator/metaline_builder.cpp

```diff
@@ -0,0 +1,217 @@
+      else if (!m_oneway)
+        Reverse();
+      else
+        return false;
+    }
+    if (m_end == line.m_start)
+    {
+      m_ways.insert(m_ways.end(), line.m_ways.begin(), line.m_ways.end());
+      m_end = line.m_end;
+      m_oneway = m_oneway || line.m_oneway;
+    }
+    else if (m_start == line.m_end)
+    {
+      m_ways.insert(m_ways.begin(), line.m_ways.begin(), line.m_ways.end());
+      m_start = line.m_start;
+      m_oneway = m_oneway || line.m_oneway;
+    }
+    else
+    {
+      return false;
+    }
+    return true;
+  }
+};
+}  // namespace
+
+namespace feature
+{
+class Segments
+{
+  std::list<LineString> m_parts;
+
+public:
+  explicit Segments(OsmElement const & way) { m_parts.push_back(LineString(way)); }
+
+  void Add(OsmElement const & way)
+  {
+    LineString line(way);
+    auto found = m_parts.end();
+    for (auto i = m_parts.begin(); i != m_parts.end(); ++i)
+    {
+      if (i->Add(line))
+      {
+        found = i;
+        break;
+      }
+    }
+    if (found == m_parts.cend())
+    {
+      m_parts.push_back(line);
+      return;
+    }
+    for (LineString & part : m_parts)
+    {
+      if (part.Add(*found))
+      {
+        m_parts.erase(found);
+        break;
+      }
+    }
+  }
+
+  std::vector<Ways> GetLongWays() const
+  {
+    std::vector<Ways> result;
+    for (LineString const & line : m_parts)
+      if (line.Ways().size() > 1)
+        result.push_back(line.Ways());
+    return result;
+  }
+};
+
+void MetalinesBuilder::operator()(OsmElement const & el, FeatureParams const & params)
+{
+  static uint32_t const highwayType = classif().GetTypeByPath({"highway"});
+  if (params.FindType(highwayType, 1) == ftype::GetEmptyValue() ||
+      el.Nodes().front() == el.Nodes().back())
+    return;
+
+  std::string name;
+  params.name.GetString(0, name);
+  if (name.empty() || params.ref.empty())
+    return;
+
+  std::size_t const key = std::hash<std::string>{}(name + '\0' + params.ref);
+  auto segment = m_data.find(key);
+  if (segment == m_data.cend())
+    m_data.emplace(key, std::make_shared<Segments>(el));
+  else
+    segment->second->Add(el);
+}
+
+void MetalinesBuilder::Flush()
+{
+  uint32_t count = 0;
+  FileWriter writer(m_filePath);
+  for (auto const & seg : m_data)
+  {
+    auto const & longWays = seg.second->GetLongWays();
+    for (Ways const & ways : longWays)
+    {
+      uint16_t size = ways.size();
+      WriteToSink(writer, size);
+      for (std::int32_t const way : ways)
+        WriteToSink(writer, way);
+      ++count;
+    }
+  }
+  LOG_SHORT(LINFO, ("Wrote", count, "metalines with OSM IDs for the entire planet to", m_filePath));
+}
+
+bool WriteMetalinesSection(std::string const & mwmPath, std::string const & metalinesPath,
+                           std::string const & osmIdsToFeatureIdsPath)
+{
+  map<osm::Id, uint32_t> osmIdToFeatureId;
+  if (!routing::ParseOsmIdToFeatureIdMapping(osmIdsToFeatureIdsPath, osmIdToFeatureId))
+    return false;
+
```

Comment:
A lot of them do.
Here is an example: https://github.com/mapsme/omim/blob/master/routing/index_graph_serialization.hpp

And it is much harder to migrate for those that don't, like geom. index.